web/htdocs/assets/js/codesearch_ui.js

```diff
@@ -45,10 +47,84 @@ function shorten(ref) {
   return ref;
 }
 
-var Match = Backbone.Model.extend({
+var MatchView = Backbone.View.extend({
+  tagName: 'div',
   initialize: function() {
+    this.model.on('change', this.render, this);
   },
-  url: function() {
+  render: function() {
+    var div = this._render();
+    this.setElement(div);
+    return this;
+  },
+  _renderLno: function(n, isMatch) {
+    return h.a({cls: 'lno', href: this.model.url(n)}, [n.toString(), isMatch ? ":" : "-"]);
+  },
+  _render: function() {
+    var i;
+    var ctx_before = [], ctx_after = [];
+    var lno = this.model.get('lno');
+    var ctxBefore = this.model.get('context_before'), clip_before = this.model.get('clip_before');
+    var ctxAfter = this.model.get('context_after'), clip_after = this.model.get('clip_after');
+
+    var lines_to_display_before = Math.max(0, ctxBefore.length - (clip_before || 0));
+    for (i = 0; i < lines_to_display_before; i ++) {
+      ctx_before.unshift(h.div([
+                                 this._renderLno(lno - i - 1, false),
+                                 this.model.get('context_before')[i]
+                               ]));
+    }
+    var lines_to_display_after = Math.max(0, ctxAfter.length - (clip_after || 0));
+    for (i = 0; i < lines_to_display_after; i ++) {
+      ctx_after.push(h.div([
+                             this._renderLno(lno + i + 1, false),
+                             this.model.get('context_after')[i]
+                           ]));
+    }
+    var line = this.model.get('line');
+    var bounds = this.model.get('bounds');
+    var pieces = [line.substring(0, bounds[0]),
+                  line.substring(bounds[0], bounds[1]),
+                  line.substring(bounds[1])];
+
+    var classes = ['match'];
+    if(clip_before !== undefined) classes.push('clip-before');
+    if(clip_after !== undefined) classes.push('clip-after');
+
+    var matchElement = h.div({cls: classes.join(' ')}, [
```

Comment:
Yikes. I can't even figure out how to make IE's `-ms-grid` layout do the right thing here. I'm not sure whether to abandon IE/Edge, or switch back to a table-based layout, where pasting as formatted text is weird.

Github uses a table, here's what happens when pasting into Gmail on Chrome:
![screenshot 2017-05-26 14 51 38](https://cloud.githubusercontent.com/assets/376107/26513875/eb0770e0-4222-11e7-8edc-485b775de9ab.png)

Perhaps I should give up on the idea of doing better, since Github are professionals and this is the best they've got (and I'm clearly not a web expert).